### ISSUE TITLE
2D displaying problems after affine transformation

### ISSUE THREAD

**Issue by Chris-N-K** [napari/napari#2616]
## 🐛 Bug

I try to use affine transformation to display my images in correct orientation to each other. When I apply a affine transformation via the affine argument the initial dimension order is displayed correctly. For all other dimension orders (roll through dimension order button) the affine transformation is not apllied correctly. This problem only happens in 2D mode, not in 3D. 

On every slice in 2D mode a warning is displayed:<br>
Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.

I suppose napari is at the current state not able to correctly operate with my affine matrix?
I'm not sure if this is a bug, simply not yet fully implemented or a failure on my part.

Correctly displayed:
![error_report_1](https://user-images.githubusercontent.com/56394171/116089944-a4d16b00-a6a3-11eb-999a-2291985aff4b.png)

Wrong:
![error_report_2](https://user-images.githubusercontent.com/56394171/116089500-3ab8c600-a6a3-11eb-8bd9-ff28157bc688.png)
![error_report_3](https://user-images.githubusercontent.com/56394171/116089508-3c828980-a6a3-11eb-8735-44e3513112a9.png)

## To Reproduce

1. Load image with affine matrix:
 ```
[[-1.17185862e+00 -4.84692848e-04  5.79785848e-03  1.69048981e+02]
 [-6.19494232e-03  1.13797526e-01 -1.09471981e+00  3.67642326e+01]
 [ 1.17442619e-04  1.16633653e+00  1.06812411e-01 -1.66103561e+02]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
![error_report_1](https://user-images.githubusercontent.com/56394171/116089130-d138b780-a6a2-11eb-90e5-8dc4e0d727a1.png)
![error_report_2](https://user-images.githubusercontent.com/56394171/116089151-d72e9880-a6a2-11eb-90eb-b636e94c4def.png)
![error_report_3](https://user-images.githubusercontent.com/56394171/116089170-dc8be300-a6a2-11eb-8cbb-f7fa2ce0be67.png)
```

2. Stay in 2D mode
3. Roll through dimension order with the "Roll dimensions order for display" button
 
## Expected behavior
Correctly display the slices in 2D

## Environment
napari: 0.4.7
Platform: Linux-3.10.0-1127.19.1.el7.x86_64-x86_64-with-glibc2.10
Python: 3.8.5 (default, Sep 4 2020, 07:30:14) [GCC 7.3.0]
Qt: 5.15.2
PyQt5: 5.15.2
NumPy: 1.20.1
SciPy: 1.6.1
Dask: 2021.03.0
VisPy: 0.6.6

OpenGL:
- GL version: 2.1 Mesa 18.3.4
- MAX_TEXTURE_SIZE: 8192

Screens:
- screen 1: resolution 1920x1080, scale 1.0

Plugins:
- console: 0.0.3
- svg: 0.1.4

**Comment by sofroniewn** [comment]
Hi @ch-n I havn't had a chance to run your example yet, but to give you a sense of the current limitation of napari, if you look at this schematic below 

![image](https://user-images.githubusercontent.com/6531703/116269322-faa12400-a732-11eb-8cfb-fc5fe782e858.png)

if you have a rotated cube and then try to take a 2D slice (the dashed line) through it then we actually can't do that right now. It's on our list to support this, but we're not there yet. Is this what you're seeing?

Note that 3D rendering is fine.

Happy to go into more details about why we have this limitation now and how we might go about improving things

**Comment by Chris-N-K** [comment]
Yes I think this is exactly the problem. 3D is fine. I guess non orthogonal slicing through an array is far from trivial. Might there be a workaround to display the image in 2D without the sheer and out of slice rotation? I try to improve the nibabel napari plug in. Therefore, I hope to include affine transformation, but If the transformation breaks the 2D view it is a problem.

**Comment by Chris-N-K** [comment]
Are there any developments on this?

**Comment by jni** [comment]
It's on our [roadmap](https://github.com/orgs/napari/projects/24/views/2?pane=info) (consistent transformations) but that's about all I can say for the moment... I will personally try to put some effort into this over the summer...
